From 2c22998047f4066e72e2354bc2af70ddc71cbfc7 Mon Sep 17 00:00:00 2001
From: David Redondo <kde@david-redondo.de>
Date: Tue, 20 Oct 2026 12:16:53 +0200
Subject: [PATCH] ConfigReader: Save file modifications timestamps in UTC

A QDateTime with Qt::LocalTime will not save the timezone it was created with.
After a system time zone change it will use the new system time zone and refer
to a different point in time.
For example 2026-10-20 10:31:46.312 CEST Qt::LocalTime will become
2026-10-20 09:32:39.711 BST Qt::LocalTime.
Fixes #2226
---
 src/common/ConfigReader.cpp | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/common/ConfigReader.cpp b/src/common/ConfigReader.cpp
index 0117a8335..848f572c9 100644
--- a/src/common/ConfigReader.cpp
+++ b/src/common/ConfigReader.cpp
@@ -148,17 +148,17 @@ namespace SDDM {
         // * m_path (classic fallback /etc/sddm.conf)
 
         QStringList files;
-        QDateTime latestModificationTime = QFileInfo(m_path).lastModified();
+        QDateTime latestModificationTime = QFileInfo(m_path).lastModified().toUTC();
 
         if (!m_sysConfigDir.isEmpty()) {
             //include the configDir in modification time so we also reload on any files added/removed
             QDir dir(m_sysConfigDir);
             if (dir.exists()) {
-                latestModificationTime = std::max(latestModificationTime,  QFileInfo(m_sysConfigDir).lastModified());
+                latestModificationTime = std::max(latestModificationTime,  QFileInfo(m_sysConfigDir).lastModified().toUTC());
                 const auto dirFiles = dir.entryInfoList(QDir::Files | QDir::NoDotAndDotDot, QDir::LocaleAware);
                 for (const QFileInfo &file : dirFiles) {
                     files << (file.absoluteFilePath());
-                    latestModificationTime = std::max(latestModificationTime, file.lastModified());
+                    latestModificationTime = std::max(latestModificationTime, file.lastModified().toUTC());
                 }
             }
         }
@@ -166,11 +166,11 @@ namespace SDDM {
             //include the configDir in modification time so we also reload on any files added/removed
             QDir dir(m_configDir);
             if (dir.exists()) {
-                latestModificationTime = std::max(latestModificationTime,  QFileInfo(m_configDir).lastModified());
+                latestModificationTime = std::max(latestModificationTime,  QFileInfo(m_configDir).lastModified().toUTC());
                 const auto dirFiles = dir.entryInfoList(QDir::Files | QDir::NoDotAndDotDot, QDir::LocaleAware);
                 for (const QFileInfo &file : dirFiles) {
                     files << (file.absoluteFilePath());
-                    latestModificationTime = std::max(latestModificationTime, file.lastModified());
+                    latestModificationTime = std::max(latestModificationTime, file.lastModified().toUTC());
                 }
             }
         }
